cache/utils.go

```diff
@@ -0,0 +1,152 @@
+// Utils provide a series of helper functions.
+
+package cache
+
+import (
+	"bytes"
+	"net/http"
+	"slices"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/gophercloud/utils/openstack/clientconfig"
+	"github.com/openstack-exporter/openstack-exporter/exporters"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/prometheus/common/expfmt"
+)
+
+// CollectCache collects the MetricsFamily for required clouds and services and stores in the cache.
+func CollectCache(
+	enableExporterFunc func(
+		string, string, string, []string, string, bool, bool, bool, bool, string, func() (string, error), log.Logger,
+	) (*exporters.OpenStackExporter, error),
+	multiCloud bool,
+	services map[string]*bool, prefix,
+	cloud string,
+	disabledMetrics []string,
+	endpointType string,
+	collectTime bool,
+	disableSlowMetrics bool,
+	disableDeprecatedMetrics bool,
+	disableCinderAgentUUID bool,
+	domainID string,
+	uuidGenFunc func() (string, error),
+	logger log.Logger,
+) error {
+	level.Info(logger).Log("msg", "Run collect cache job")
+	cacheBackend := GetCache()
+
+	clouds := []string{}
+
+	if multiCloud {
+		cloudsConfig, err := clientconfig.LoadCloudsYAML()
+		if err != nil {
+			return err
+		}
+		for cloud := range cloudsConfig {
+			clouds = append(clouds, cloud)
+		}
+	}
+	if cloud != "" && !multiCloud {
+		clouds = append(clouds, cloud)
+	}
+
+	enabledServices := []string{}
+	for service, disabled := range services {
+		if !*disabled {
+			enabledServices = append(enabledServices, service)
+		}
+	}
+
+	for _, cloud := range clouds {
+		level.Info(logger).Log("msg", "Start update cache data", "cloud", cloud)
+		// Update cloud's cache once finish all exporters' collection job. so we won't mix the old
+		// and new metrics in the cache and confuse users.
+		cloudCache := NewCloudCache()
+
+		for _, service := range enabledServices {
+			level.Info(logger).Log("msg", "Start collect cache data", "cloud", cloud, "service", service)
+			exp, err := enableExporterFunc(service, prefix, cloud, disabledMetrics, endpointType, collectTime, disableSlowMetrics, disableDeprecatedMetrics, disableCinderAgentUUID, domainID, nil, logger)
```

Comment:
Since you imported the `exporters`, can you just use `exporters.EnableExporter` instead of the anonymous `enableExporterFunc` argument ?

Reply:
It's a practice to inject the func so you can more easy to write the unit test.

See https://github.com/jneo8/openstack-exporter/blob/fb699a36e0eba85590b5a311d467259766d9a8a3/cache/utils_test.go#L88

Reply:
Okay, understood. but not sure about that: should we make the actual code better or unit test better. (not an expert in Go). Feel free to resolve because I can't